pypeclub/OpenPype
# General: Inclusive handles

## Brief description
Added ability to have inclusive handles per project.

## Description
Inclusive handles changes what `frameStart` and `frameEnd` stored on entities represents. With inclusive handles they represent full range. Even when handles should not be used we need to know their value to be able know frame range.

| | Frame Start | Frame End | Handle Start | Handle End | Full range | Range without handles |
| --- | --- | --- | --- | --- | --- | --- |
| Extend handles | 1001 | 1100 | 10 | 10 | 991-1110 | 1001-1100 |
| Inclusive handles | 1001 | 1100 | 10 | 10 | 1001-1100 | 1011-1090 |

It requires a lot of changes when data about frames are loaded from asset and version because they need recalculations.

To keep the code simpler it was created a unified frame info which is used when frame data are loaded from asset or version document. Based on anatomy attribute `handlesExtend` is recalculating frame ranges. This allows to keep current code unchanded in most of cases as inclusive handles are converted into extending.

I hope I've changed all publishing and loading code where frame start and frame end are used when reading from asset and version document.

New Anatomy key `handlesExtend` was also added to ftrack custom attributes.

## TODOs
Editorial publishing was not modified. I couldn't find out how handles are handles handled in that case. I know that editorial can create hierarchy (asset documents) where it should be stored the way project expect it so I'll need @jakubjezek001 help.

It is not properly validated if the frame range is valid.

## Notes
Changing the `handlesExtend` during production would probably break already existing version documents.

## Testing notes
1. Change `handlesExtend` to `False` on a project.
    - since that moment all frames should be recalculated
2. Check if frames are calculated in right way.

If `handlesExtend` is set to `True` all should work as before.
